### ISSUE TITLE
Saving Catalogs will delete all ILocalized strings 

### ISSUE THREAD

**Issue by Farshidgolkarihagh** [nopSolutions/nopCommerce#5700]
nopCommerce version: 4.4

Steps to reproduce the problem: 

- Add 2 language to nop commerce
- Create a category
- Add some text into both languages
- Edit the catalog and modify the content for the 2nd language
- Save and continue catalog

Actual

- Page will load but the content of the 2nd language modified is wiped 
- Reloading the page will display the new values 

Expected

- On page save, it should save the new data and fetch the data on page load.
- When I tracked the problem seems like we have a problem with the 
public virtual async Task<IList<T>> PrepareLocalizedModelsAsync<T>(Action<T, int> configure = null) where T : ILocalizedLocaleModel

Have a look at the example below: 

    class Program
    {
        static async Task Main(string[] args)
        {
            Action<TestMe> dsf = async (i) => { await TestOne(i);  };
            var oo = new TestMe { Value = "Original" };
            Console.WriteLine($"Test1 before chanege: { oo.Value}");
            dsf.Invoke(oo);
            Console.WriteLine($"Test1 Final Value: { oo.Value}");

            Func<TestMe, string, Task<TestMe>> test2 = async (i, o) => { return await TestTwo(i); };
            var t2 = new TestMe { Value = "Original" };
            Console.WriteLine($"Test2 before chanege: { oo.Value}");
            await test2(t2, "");
            Console.WriteLine($"Test2 Final Value: { t2.Value}");
            Console.WriteLine($"Waiting to read key");
            Console.ReadLine();
        }

        private static async Task TestOne(TestMe taskName)
        {
            await Task.Delay(1000);
            taskName.Value = "Modified";
        }
        private static async Task<TestMe> TestTwo(TestMe taskName)
        {
            await Task.Delay(1000);
            taskName.Value = "Modified";
            return taskName;
        }
    }

    public class TestMe
    {
        public string Value { get; set; }
    }

When you run the console application you will see that the value has not been modified when we invoke a function that changed string. I suspect when performing both saving and loading of the data this bug is coming to surface.

**Comment by skoshelev** [comment]
Hi @Farshidgolkarihagh. I could not repeat this problem on the latest source codes, please specify the minor version on which you encountered this problem, was it version 4.40.3?

**Comment by Farshidgolkarihagh** [comment]
I am using version 4.40.3. I can provide you with a test enviroment and access to the code that I have compiled OR username and password to the test enviroment, if you can message me your emails address.

**Comment by skoshelev** [comment]
Hi @Farshidgolkarihagh. If possible, record a screen capture showing the problem. You can send data to access the test environment to sergey.k@nopcommerce.com. We will be glad to receive any additional information on the problem.

**Comment by RomanovM** [comment]
See also [this topic](https://www.nopcommerce.com/boards/topic/90614/4402-product-edit-using-multiple-language) on our forums, could this be related?

**Comment by Farshidgolkarihagh** [comment]
Yes it the exact same issue, I had this problem all over the product and had to do a full fix in all places where PrepareLocalizedModelsAsync is used. I will post a screen share tomorrow.

**Comment by Farshidgolkarihagh** [comment]
I will post my fix here tomorrow

**Comment by AndreiMaz** [comment]
@Farshidgolkarihagh have you had a chance to prepare it?

**Comment by Farshidgolkarihagh** [comment]
Hi

Sorry for the delay in getting back. I will follow up tomorrow. Has anyone
else had the same issue?

Kind regards,
Farshid

This email is sent from my pixel so apologies for any typos

**Comment by Farshidgolkarihagh** [comment]
Hi 

I have sent you an email with the username and password to a test enviroment. 

Kind regards,
Farshid

**Comment by skoshelev** [comment]
We have not yet been able to reproduce the problem, we think that it lies either in the specific structure of the tables or in the conflict of third-party plugins. At the moment, we are closing the ticket, if additional data appears, we will return to its analysis.

**Comment by brakim** [comment]
I had the exakt same problem and noticed it somehow relate to LoadAllLocalizedPropertiesOnStartup and having about 20k records in LocalizedProperty.

Turning off LoadAllLocalizedPropertiesOnStartup fixed the problem.

**Comment by Ban3ta** [comment]
> I had the exakt same problem and noticed it somehow relate to LoadAllLocalizedPropertiesOnStartup and having about 20k records in LocalizedProperty.
> 
> Turning off LoadAllLocalizedPropertiesOnStartup fixed the problem.

Thank you mate, this worked for me

**Comment by bellamatte** [comment]
Hi, I had the same problem (noticed while editing products).
Actually turning off LoadAllLocalizedPropertiesOnStartup seems to solve it, but I think it's some kind of coincidence.
In fact, I maybe found the cause.

The method `LocalizedModelFactory.PrepareLocalizedModelsAsync` accepts an Action<T, int> as argument.
It invokes that action this way:

```
                //invoke the model configuration action
                configure?.Invoke(localizedModel, localizedModel.LanguageId);
```

But in the method `ProductModelFactory.PrepareProductModelAsync` the same action (defined as Action<T, int>) is assigned this way:

```
                //define localized model configuration action
                localizedModelConfiguration = async (locale, languageId) =>
                {
                    locale.Name = await _localizationService.GetLocalizedAsync(product, entity => entity.Name, languageId, false, false);
                    locale.FullDescription = await _localizationService.GetLocalizedAsync(product, entity => entity.FullDescription, languageId, false, false);
                    locale.ShortDescription = await _localizationService.GetLocalizedAsync(product, entity => entity.ShortDescription, languageId, false, false);
                    locale.MetaKeywords = await _localizationService.GetLocalizedAsync(product, entity => entity.MetaKeywords, languageId, false, false);
                    locale.MetaDescription = await _localizationService.GetLocalizedAsync(product, entity => entity.MetaDescription, languageId, false, false);
                    locale.MetaTitle = await _localizationService.GetLocalizedAsync(product, entity => entity.MetaTitle, languageId, false, false);
                    locale.SeName = await _urlRecordService.GetSeNameAsync(product, languageId, false, false);
                };
```

As you can see, localizedModelConfiguration is defined as Action, but an async function is assigned to it.
So, when it's invoked in `PrepareLocalizedModelsAsync` there is no guarantee the invocation finishes.
So I think that the page is rendered before this "action" is completed, when the localized property was null.

When it's called just after a Save, the cache has just been cleared, so if LoadAllLocalizedPropertiesOnStartup is on, it needs to reload all the localized properties. Since it takes a little bit longer, it can't finish the task before the page rendering.

SOLUTION:

Since all the occurrences seem to be async, they need to be turned into Func<T, int, Task>, so it can be awaited.

The method will end up like that:

```
        public virtual async Task<IList<T>> PrepareLocalizedModelsAsync<T>(Func<T, int, Task> configure = null) where T : ILocalizedLocaleModel
        {
            //get all available languages
            var availableLanguages = await _languageService.GetAllLanguagesAsync(true);

            //prepare models
            var localizedModels = await availableLanguages.SelectAwait(async language =>
            {
                //create localized model
                var localizedModel = Activator.CreateInstance<T>();

                //set language
                localizedModel.LanguageId = language.Id;

                //invoke the model configuration action
                if (configure != null)
                    await configure(localizedModel, localizedModel.LanguageId);

                return localizedModel;
            }).ToListAsync();

            return localizedModels;
        }
```

**Comment by Farshidgolkarihagh** [comment]
I made this suggestion in my first post too:

Since all the occurrences are async, they need to be turned into Func<T, int, Task>, so they can be awaited.

**Comment by skoshelev** [comment]
Hi @bellamatte. Thank you for your help and investigation, after detail analiz we made the proposed changing

Closed #5700
